p4-16/spec/P4-16-spec.mdk

```diff
@@ -7716,33 +7713,51 @@ The evaluation of a P4 program is done in two stages:
-- Expressions evaluating to a list type, where all elements are
-  compile-time known values.
-- Instances constructed by instance declarations (Section
-  [#sec-instantiations]) and constructor invocations.
-- A legal cast applied to a compile-time known value
-- The following expressions (`+`, `-`, `*`, `/ `, `%`, `!`, `&`, `|`, `&&`, `||`, `<< `, `>> `, `~` `, `\ `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`)
-  when their operands are all compile-time known values.
+- The expression `{#}` representing an invalid header or header union value.
+- Instances constructed by instance declarations (Section [#sec-instantiations]) and constructor invocations.
+- Identifiers that represent declared types, actions, tables, parsers, controls, or packages.
+- Tuple expression where all components are local compile-time known values.
+- Structure-valued expressions, where all fields are local compile-time known values.
+- Expressions evaluating to a list type, where all elements are local compile-time known values.
+- Legal casts applied to local compile-time known values.
+- The following expressions (`+`, `-`, `|+|`, `|-|`, `*`, `/ `, `%`, `!`, `&`, `|`, `^`, `&&`, `||`, `<< `, `>>`, `~`, `/`, `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`) when their operands are all local compile-time known values.
+- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`, `e.maxSizeInBits()` and `e.maxSizeInBytes()` where the type of `e` is not generic.
+
+The following are compile-time known values:
+
+- All local compile-time known values.
+- Constructor parameters.
 - Identifiers declared as constants using the `const` keyword.
-- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`,
-  `e.maxSizeInBits()` and `e.maxSizeInBytes()`
+- Tuple expression where all components are compile-time known values.
+- Expressions evaluating to a list type, where all elements are compile-time known values.
+- Structure-valued expressions, where all fields are compile-time known values.
+- Expressions evaluating to a list type, where all elements are compile-time known values.
+- Legal casts applied to compile-time known values.
+- The following expressions (`+`, `-`, `|+|`, `|-|`, `*`, `/ `, `%`, `cast`, `!`, `&`, `|`, `^`, `&&`, `||`, `<< `, `>> `, `~`, `/`, `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`) when their operands are all compile-time known values.
+- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`, `e.maxSizeInBits()` and `e.maxSizeInBytes()` where the the type of `e` is generic.
+
+Intuitively, the main difference between _compile-time known values_
+and _local compile-time known values_ is that the former also contains
+constructor parameters. The distinction is important when it comes to
+defining the meaning of features like types. For example, in the type
+`bit<e>`, the expression `e` must be a local compile-time known value
+as the value of a constructor parameter, is only determined when the
```

Comment:
I think something got lost here? 

"For example, in the type `bit<e>`, the expression `e` must be a local compile-time known value" makes sense

" as the value of a constructor parameter, is only determined when the
constructor is applied to a value." was maybe supposed to be up where it says "the former also contains constructor parameters"?